File: src/parser.py
"""
数据解析层 - 将 API 响应转换为数据模型
"""
from abc import ABC, abstractmethod
from datetime import datetime
from typing import List, Optional

from models import (
    VideoRecord,
    LiveRecord,
    ArticleRecord,
    ContentType,
    FollowingRecord,
    FavFolderRecord,
    FavResourceRecord,
)
from logger import logger


class BaseParser(ABC):
    """解析器基类"""

    @abstractmethod
    def parse(self, item: dict):
        """解析单个项目"""
        pass

    @staticmethod
    def parse_timestamp(timestamp: int) -> datetime:
        """解析时间戳"""
        if timestamp:
            return datetime.fromtimestamp(timestamp)
        return datetime.now()

    @staticmethod
    def get_history(item: dict) -> dict:
        """获取游标接口的 history 嵌套对象"""
        return item.get("history") or {}


class VideoParser(BaseParser):
    """视频解析器"""

    def parse(self, item: dict) -> Optional[VideoRecord]:
        """解析视频记录"""
        history = self.get_history(item)
        bvid = history.get("bvid") or item.get("bvid", "")
        if not bvid:
            return None

        title = item.get("title", "未知标题")
        duration = item.get("duration", 0)
        progress = item.get("progress", 0)
        if progress == -1:  # 游标接口中 -1 表示已看完
            progress = duration
        view_at_ts = item.get("view_at", 0)

        # 获取 UP 主信息
        author = self._get_author(item)
        category = self._get_category(item)

        return VideoRecord(
            title=title,
            bvid=bvid,
            author=author,
            view_time=self.parse_timestamp(view_at_ts),
            duration=duration,
            progress=progress,
            link=f"https://www.bilibili.com/video/{bvid}",
            category=category
        )

    @staticmethod
    def _get_category(item: dict) -> str:
        """获取视频分区/分类"""
        # 方式1: 游标接口常见字段 tname / tag_name / typename
        for key in ("tname", "tag_name", "typename"):
            value = item.get(key)
            if value:
                return value

        # 方式2: tag 列表取第一个
        tags = item.get("tag")
        if isinstance(tags, list) and tags:
            return tags[0].get("tag_name") or tags[0].get("name") or ""

        # 方式3: 嵌套 obj 中的分类名
        obj = item.get("stat") or item.get("dynamic") or {}
        if isinstance(obj, dict):
            return obj.get("tname") or ""

        return ""

    @staticmethod
    def _get_author(item: dict) -> str:
        """获取 UP 主名称"""
        # 方式1: 游标接口的 author_name 字段
        if item.get("author_name"):
            return item["author_name"]

        # 方式2: upper 字段
        upper = item.get("upper", {})
        if upper:
            return upper.get("name") or upper.get("uname")

        # 方式3: owner 字段
        owner = item.get("owner", {})
        if owner:
            return owner.get("name") or owner.get("uname")

        # 方式4: 根对象 name 字段
        return item.get("name") or "未知UP主"


class LiveParser(BaseParser):
    """直播解析器"""

    def parse(self, item: dict) -> Optional[LiveRecord]:
        """解析直播记录"""
        history = self.get_history(item)
        room_id = history.get("oid") or item.get("room_id", "")
        if not room_id:
            return None

        title = item.get("title", "未知标题")
        view_at_ts = item.get("view_at", 0)

        # 获取主播信息
        upper = item.get("upper") or item.get("owner") or {}
        author = item.get("author_name") or upper.get("name") or upper.get("uname", "未知主播")

        # badge 在直播场景通常表示 "直播中"/"未开播" 等状态
        category = item.get("badge") or ""

        return LiveRecord(
            title=title,
            room_id=str(room_id),
            author=author,
            view_time=self.parse_timestamp(view_at_ts),
            link=f"https://live.bilibili.com/{room_id}",
            category=category
        )


class ArticleParser(BaseParser):
    """专栏解析器"""

    def parse(self, item: dict) -> Optional[ArticleRecord]:
        """解析专栏记录"""
        history = self.get_history(item)
        article_id = history.get("oid") or item.get("id", "")
        if not article_id:
            return None

        title = item.get("title", "未知标题")
        view_at_ts = item.get("view_at", 0)

        # 获取作者信息
        upper = item.get("upper") or item.get("author") or {}
        author = item.get("author_name") or upper.get("name") or upper.get("uname", "未知作者")

        # 专栏分类：优先 category，否则按模板 ID 兜底为常见分区
        category = item.get("category") or ""
        if not category:
            category = self._category_from_template(item.get("template_id"))

        return ArticleRecord(
            title=title,
            article_id=str(article_id),
            author=author,
            view_time=self.parse_timestamp(view_at_ts),
            link=f"https://www.bilibili.com/read/cv{article_id}",
            category=category
        )

    @staticmethod
    def _category_from_template(template_id) -> str:
        """根据专栏模板 ID 映射常见分类（B站模板 ID 多为 4 位数字）"""
        mapping = {
            "4": "动画",
            "17": "游戏",
            "3": "音乐",
            "129": "舞蹈",
            "36": "知识",
            "188": "数码",
            "95": "科技",
            "122": "生活",
            "160": "时尚",
            "211": "美食",
            "119": "鬼畜",
            "155": "娱乐",
            "5": "影视",
            "181": "影视",
            "202": "资讯",
        }
        tid = str(template_id) if template_id is not None else ""
        return mapping.get(tid, "")


class UserInfoParser(BaseParser):
    """用户信息解析器（注册时间等）"""

    def parse(self, item: dict) -> Optional[datetime]:
        """解析用户卡片数据，返回注册时间"""
        return self.parse_registration_time(item)

    def parse_registration_time(self, raw_data: dict) -> Optional[datetime]:
        """从用户卡片 API 响应中解析注册时间

        Args:
            raw_data: API 原始响应

        Returns:
            注册时间 datetime，失败返回 None
        """
        if not isinstance(raw_data, dict):
            return None

        reg_ts = raw_data.get("data", {}).get("card", {}).get("regtime")
        if reg_ts:
            return self.parse_timestamp(int(reg_ts))
        return None


class HistoryParser:
    """历史记录解析器 - 统一调度"""

    def __init__(self):
        self.parsers = {
            "archive": VideoParser(),
            "live": LiveParser(),
            "article": ArticleParser()
        }

    def parse(self, raw_data: dict) -> List:
        """解析 API 返回的历史数据"""
        records = []

        # 获取列表数据
        history_list = raw_data.get("data", {}).get("list", []) if isinstance(raw_data, dict) else raw_data

        for item in history_list:
            # 游标接口 business 位于 history 嵌套对象中，兼容旧版扁平字段
            business_type = (item.get("history") or {}).get("business") or item.get("business")
            parser = self.parsers.get(business_type)

            if parser:
                record = parser.parse(item)
                if record:
                    records.append(record)

        return records

    def parse_page(self, raw_data: dict) -> tuple:
        """解析一页数据，返回 (记录列表, 是否还有更多)"""
        records = self.parse(raw_data)
        cursor = raw_data.get("data", {}).get("cursor", {}) if isinstance(raw_data, dict) else {}
        has_more = bool(cursor.get("max") or cursor.get("view_at"))
        return records, has_more


class FollowingParser:
    """关注列表解析器"""

    def parse(self, raw_data: dict) -> List[FollowingRecord]:
        items = raw_data.get("data", {}).get("list", []) if isinstance(raw_data, dict) else raw_data
        records = []
        for item in items:
            mid = item.get("mid")
            if not mid:
                continue
            official = item.get("official_verify") or {}
            records.append(FollowingRecord(
                mid=str(mid),
                name=item.get("uname") or item.get("username") or "未知UP主",
                face=item.get("face", ""),
                sign=item.get("sign", ""),
                official=official.get("desc", "") if isinstance(official, dict) else str(official),
                level=item.get("level", 0),
                link=f"https://space.bilibili.com/{mid}"
            ))
        return records


class FavFolderParser:
    """收藏夹文件夹解析器"""

    def parse(self, raw_data: dict) -> List[FavFolderRecord]:
        items = raw_data.get("data", {}).get("list", []) if isinstance(raw_data, dict) else raw_data
        records = []
        for item in items:
            folder_id = item.get("id")
            if not folder_id:
                continue
            records.append(FavFolderRecord(
                folder_id=str(folder_id),
                title=item.get("title", "未命名收藏夹"),
                media_count=item.get("media_count", 0),
                link=f"https://space.bilibili.com/favlist?fid={folder_id}"
            ))
        return records


class FavResourceParser:
    """收藏夹内容解析器"""

    def parse(self, raw_data: dict) -> List[FavResourceRecord]:
        items = raw_data.get("data", {}).get("medias", []) if isinstance(raw_data, dict) else raw_data
        records = []
        for item in items:
            title = item.get("title", "未知标题")
            upper = item.get("upper", {}) or {}
            bvid = item.get("bvid") or item.get("id", "")
            link = f"https://www.bilibili.com/video/{bvid}" if bvid else ""
            records.append(FavResourceRecord(
                title=title,
                bvid=bvid,
                link=link,
                author=upper.get("name") or upper.get("uname", ""),
                cover=item.get("cover", ""),
                resource_type=self._type_name(item.get("type", 2))
            ))
        return records

    @staticmethod
    def _type_name(type_id: int) -> str:
        mapping = {2: "video", 12: "article", 21: "video", 22: "bangumi"}
        return mapping.get(type_id, "video")

File: src/core/parser.cpp
#include "parser.h"

#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonValue>

namespace bili {

namespace {

qint64 toLongLong(const QJsonValue& value) {
    if (value.isDouble()) {
        return static_cast<qint64>(value.toDouble());
    }
    if (value.isString()) {
        return value.toString().toLongLong();
    }
    return 0;
}

QString toString(const QJsonValue& value) {
    if (value.isString()) return value.toString();
    if (value.isDouble()) return QString::number(value.toDouble(), 'f', 0);
    return QString();
}

} // namespace

HistoryParseResult Parser::parseHistory(const QJsonObject& data) {
    HistoryParseResult result;
    const QJsonObject dataObj = data.value(QStringLiteral("data")).toObject();
    const QJsonArray list = dataObj.value(QStringLiteral("list")).toArray();

    for (const QJsonValue& value : list) {
        if (!value.isObject()) continue;
        RecordPtr record = parseHistoryItem(value.toObject());
        if (record) {
            result.records.push_back(record);
        }
    }

    const QJsonObject cursor = dataObj.value(QStringLiteral("cursor")).toObject();
    result.pagination.hasMore = cursor.contains(QStringLiteral("max")) ||
                                cursor.contains(QStringLiteral("view_at"));
    result.pagination.nextCursor = cursor.value(QStringLiteral("max")).toVariant().toString();
    result.pagination.pageSize = cursor.value(QStringLiteral("ps")).toInt(20);
    result.pagination.total = dataObj.value(QStringLiteral("page")).toObject().value(QStringLiteral("total")).toInt();

    return result;
}

RecordPtr Parser::parseHistoryItem(const QJsonObject& item) {
    const QJsonObject history = item.value(QStringLiteral("history")).toObject();
    const QString business = history.value(QStringLiteral("business")).toString()
                                 .toLower();

    if (business == QStringLiteral("archive")) {
        return parseVideo(item, history);
    }
    if (business == QStringLiteral("live")) {
        return parseLive(item, history);
    }
    if (business == QStringLiteral("article")) {
        return parseArticle(item, history);
    }

    // 兼容扁平字段
    const QString flatBusiness = item.value(QStringLiteral("business")).toString().toLower();
    if (flatBusiness == QStringLiteral("archive")) return parseVideo(item, history);
    if (flatBusiness == QStringLiteral("live")) return parseLive(item, history);
    if (flatBusiness == QStringLiteral("article")) return parseArticle(item, history);

    return nullptr;
}

RecordPtr Parser::parseVideo(const QJsonObject& item, const QJsonObject& history) {
    const QString bvid = history.value(QStringLiteral("bvid")).toString().isEmpty()
                             ? item.value(QStringLiteral("bvid")).toString()
                             : history.value(QStringLiteral("bvid")).toString();
    if (bvid.isEmpty()) {
        return nullptr;
    }

    auto record = std::make_shared<VideoRecord>();
    record->type = RecordType::Video;
    record->bvId = bvid;
    record->bvid = bvid;
    record->title = item.value(QStringLiteral("title")).toString();
    record->authorName = extractAuthorName(item);
    record->authorId = toLongLong(item.value(QStringLiteral("mid")));
    record->category = extractVideoCategory(item);
    record->duration = toLongLong(item.value(QStringLiteral("duration")));

    int progress = item.value(QStringLiteral("progress")).toInt(-2);
    if (progress == -1) {
        progress = record->duration;
    }
    record->progressPercent = record->duration > 0
                                  ? static_cast<int>((progress * 100.0) / record->duration)
                                  : 0;
    record->progress = QStringLiteral("%1%").arg(record->progressPercent);
    record->viewAt = parseTimestamp(toLongLong(item.value(QStringLiteral("view_at"))));
    record->coverUrl = item.value(QStringLiteral("cover")).toString();
    record->cid = toLongLong(item.value(QStringLiteral("cid")));

    QJsonDocument doc(item);
    record->rawJson = QString::fromUtf8(doc.toJson(QJsonDocument::Compact));

    return record;
}

RecordPtr Parser::parseLive(const QJsonObject& item, const QJsonObject& history) {
    const QString roomId = history.value(QStringLiteral("oid")).toString().isEmpty()
                               ? item.value(QStringLiteral("room_id")).toString()
                               : history.value(QStringLiteral("oid")).toString();
    if (roomId.isEmpty()) {
        return nullptr;
    }

    auto record = std::make_shared<LiveRecord>();
    record->type = RecordType::Live;
    record->roomId = roomId;
    record->bvid = roomId;
    record->title = item.value(QStringLiteral("title")).toString();
    record->authorName = extractAuthorName(item);
    record->authorId = toLongLong(item.value(QStringLiteral("mid")));
    record->category = item.value(QStringLiteral("badge")).toString();
    record->viewAt = parseTimestamp(toLongLong(item.value(QStringLiteral("view_at"))));
    record->coverUrl = item.value(QStringLiteral("cover")).toString();
    record->liveId = toLongLong(item.value(QStringLiteral("live_id")));
    record->liveStatus = item.value(QStringLiteral("live_status")).toString();

    QJsonDocument doc(item);
    record->rawJson = QString::fromUtf8(doc.toJson(QJsonDocument::Compact));

    return record;
}

RecordPtr Parser::parseArticle(const QJsonObject& item, const QJsonObject& history) {
    const QString articleId = history.value(QStringLiteral("oid")).toString().isEmpty()
                                  ? item.value(QStringLiteral("id")).toString()
                                  : history.value(QStringLiteral("oid")).toString();
    if (articleId.isEmpty()) {
        return nullptr;
    }

    auto record = std::make_shared<ArticleRecord>();
    record->type = RecordType::Article;
    record->cvId = articleId;
    record->bvid = articleId;
    record->title = item.value(QStringLiteral("title")).toString();
    record->authorName = extractAuthorName(item);
    record->authorId = toLongLong(item.value(QStringLiteral("mid")));
    record->category = item.value(QStringLiteral("category")).toString();
    if (record->category.isEmpty()) {
        record->category = articleCategoryFromTemplate(item.value(QStringLiteral("template_id")).toInt());
    }
    record->viewAt = parseTimestamp(toLongLong(item.value(QStringLiteral("view_at"))));
    record->coverUrl = item.value(QStringLiteral("cover")).toString();
    record->categoryId = toLongLong(item.value(QStringLiteral("category_id")));

    QJsonDocument doc(item);
    record->rawJson = QString::fromUtf8(doc.toJson(QJsonDocument::Compact));

    return record;
}

QString Parser::extractVideoCategory(const QJsonObject& item) {
    const QStringList keys = {QStringLiteral("tname"), QStringLiteral("tag_name"), QStringLiteral("typename")};
    for (const QString& key : keys) {
        const QString value = item.value(key).toString();
        if (!value.isEmpty()) return value;
    }

    const QJsonArray tags = item.value(QStringLiteral("tag")).toArray();
    if (!tags.isEmpty()) {
        const QJsonObject first = tags.first().toObject();
        const QString tagName = first.value(QStringLiteral("tag_name")).toString();
        if (!tagName.isEmpty()) return tagName;
        const QString name = first.value(QStringLiteral("name")).toString();
        if (!name.isEmpty()) return name;
    }

    const QJsonObject stat = item.value(QStringLiteral("stat")).toObject();
    const QString tname = stat.value(QStringLiteral("tname")).toString();
    if (!tname.isEmpty()) return tname;

    return QString();
}

QString Parser::extractAuthorName(const QJsonObject& item) {
    const QString authorName = item.value(QStringLiteral("author_name")).toString();
    if (!authorName.isEmpty()) return authorName;

    const QJsonObject upper = item.value(QStringLiteral("upper")).toObject();
    if (!upper.isEmpty()) {
        const QString name = upper.value(QStringLiteral("name")).toString();
        if (!name.isEmpty()) return name;
        return upper.value(QStringLiteral("uname")).toString();
    }

    const QJsonObject owner = item.value(QStringLiteral("owner")).toObject();
    if (!owner.isEmpty()) {
        const QString name = owner.value(QStringLiteral("name")).toString();
        if (!name.isEmpty()) return name;
        return owner.value(QStringLiteral("uname")).toString();
    }

    const QString rootName = item.value(QStringLiteral("name")).toString();
    if (!rootName.isEmpty()) return rootName;

    return QStringLiteral("未知UP主");
}

QString Parser::articleCategoryFromTemplate(int templateId) {
    static const QMap<int, QString> mapping = {
        {4, QStringLiteral("动画")},
        {17, QStringLiteral("游戏")},
        {3, QStringLiteral("音乐")},
        {129, QStringLiteral("舞蹈")},
        {36, QStringLiteral("知识")},
        {188, QStringLiteral("数码")},
        {95, QStringLiteral("科技")},
        {122, QStringLiteral("生活")},
        {160, QStringLiteral("时尚")},
        {211, QStringLiteral("美食")},
        {119, QStringLiteral("鬼畜")},
        {155, QStringLiteral("娱乐")},
        {5, QStringLiteral("影视")},
        {181, QStringLiteral("影视")},
        {202, QStringLiteral("资讯")}
    };
    return mapping.value(templateId, QString());
}

FollowingList Parser::parseFollowing(const QJsonObject& data) {
    FollowingList result;
    const QJsonObject dataObj = data.value(QStringLiteral("data")).toObject();
    const QJsonArray list = dataObj.value(QStringLiteral("list")).toArray();

    for (const QJsonValue& value : list) {
        if (!value.isObject()) continue;
        const QJsonObject item = value.toObject();
        const qint64 mid = toLongLong(item.value(QStringLiteral("mid")));
        if (mid == 0) continue;

        FollowingUser user;
        user.mid = mid;
        user.name = item.value(QStringLiteral("uname")).toString();
        if (user.name.isEmpty()) {
            user.name = item.value(QStringLiteral("username")).toString();
        }
        if (user.name.isEmpty()) {
            user.name = QStringLiteral("未知UP主");
        }
        user.faceUrl = item.value(QStringLiteral("face")).toString();
        user.sign = item.value(QStringLiteral("sign")).toString();
        user.level = item.value(QStringLiteral("level")).toInt();

        const QJsonObject official = item.value(QStringLiteral("official_verify")).toObject();
        user.officialVerify = official.value(QStringLiteral("type")).toInt();

        result.push_back(user);
    }

    return result;
}

FavoriteFolderList Parser::parseFavoriteFolders(const QJsonObject& data) {
    FavoriteFolderList result;
    const QJsonObject dataObj = data.value(QStringLiteral("data")).toObject();
    const QJsonArray list = dataObj.value(QStringLiteral("list")).toArray();

    for (const QJsonValue& value : list) {
        if (!value.isObject()) continue;
        const QJsonObject item = value.toObject();
        const qint64 id = toLongLong(item.value(QStringLiteral("id")));
        if (id == 0) continue;

        FavoriteFolder folder;
        folder.id = id;
        folder.name = item.value(QStringLiteral("title")).toString();
        if (folder.name.isEmpty()) {
            folder.name = QStringLiteral("未命名收藏夹");
        }
        folder.mediaCount = toLongLong(item.value(QStringLiteral("media_count")));
        result.push_back(folder);
    }

    return result;
}

std::vector<FavoriteItem> Parser::parseFavoriteResources(const QJsonObject& data) {
    std::vector<FavoriteItem> result;
    const QJsonObject dataObj = data.value(QStringLiteral("data")).toObject();
    const QJsonArray medias = dataObj.value(QStringLiteral("medias")).toArray();

    for (const QJsonValue& value : medias) {
        if (!value.isObject()) continue;
        const QJsonObject item = value.toObject();

        FavoriteItem fav;
        fav.id = toString(item.value(QStringLiteral("id")));
        fav.title = item.value(QStringLiteral("title")).toString();
        fav.bvid = item.value(QStringLiteral("bvid")).toString();
        fav.coverUrl = item.value(QStringLiteral("cover")).toString();
        fav.type = recordTypeFromString(favoriteTypeName(item.value(QStringLiteral("type")).toInt(2)));

        const QJsonObject upper = item.value(QStringLiteral("upper")).toObject();
        fav.upperName = upper.value(QStringLiteral("name")).toString();
        if (fav.upperName.isEmpty()) {
            fav.upperName = upper.value(QStringLiteral("uname")).toString();
        }
        fav.upperId = toLongLong(upper.value(QStringLiteral("mid")));
        fav.favTime = parseTimestamp(toLongLong(item.value(QStringLiteral("fav_time"))));

        result.push_back(fav);
    }

    return result;
}

QString Parser::favoriteTypeName(int typeId) {
    static const QMap<int, QString> mapping = {
        {2, QStringLiteral("video")},
        {12, QStringLiteral("article")},
        {21, QStringLiteral("video")},
        {22, QStringLiteral("bangumi")}
    };
    return mapping.value(typeId, QStringLiteral("video"));
}

UserInfo Parser::parseUserCard(const QJsonObject& data) {
    UserInfo info;
    const QJsonObject dataObj = data.value(QStringLiteral("data")).toObject();
    const QJsonObject card = dataObj.value(QStringLiteral("card")).toObject();

    info.mid = toLongLong(card.value(QStringLiteral("mid")));
    info.name = card.value(QStringLiteral("name")).toString();
    info.sign = card.value(QStringLiteral("sign")).toString();
    info.faceUrl = card.value(QStringLiteral("face")).toString();
    info.level = card.value(QStringLiteral("level_info")).toObject().value(QStringLiteral("current_level")).toInt();
    info.registrationTime = parseRegistrationTime(data);
    info.registrationTimeText = info.registrationTime.toString(QStringLiteral("yyyy-MM-dd"));

    return info;
}

QDateTime Parser::parseRegistrationTime(const QJsonObject& data) {
    const QJsonObject dataObj = data.value(QStringLiteral("data")).toObject();
    const QJsonObject card = dataObj.value(QStringLiteral("card")).toObject();
    const qint64 regTs = toLongLong(card.value(QStringLiteral("regtime")));
    if (regTs > 0) {
        return parseTimestamp(regTs);
    }
    return QDateTime();
}

QDateTime Parser::parseTimestamp(qint64 ts) {
    if (ts <= 0) {
        return QDateTime::currentDateTime();
    }
    if (ts > 1e12) {
        // 毫秒时间戳
        return QDateTime::fromMSecsSinceEpoch(ts);
    }
    return QDateTime::fromSecsSinceEpoch(ts);
}

} // namespace bili
